lib/id3c/api/datastore.py

```diff
@@ -382,6 +383,63 @@ def fetch_identifier_set_uses(session: DatabaseSession) -> Any:
 
         return list(cursor)
 
+@export
+@catch_permission_denied
+def store_sample(session: DatabaseSession, sample: dict) -> Any:
+    """"
+    Validate the given *sample* and insert or update in the backing database.
+
+    Returns a list of in the same order as the input, with each object including
+    the ``sample_id`` (string), ``status`` (string) to indicate if inserted,
+    updated, or validation failed, and ``details`` to indicate reason for
+    failed validation.
+    """
+    with session:
+        sample_barcode = sample.pop("sample_id", None)
+        sample_identifier = find_identifier(session, sample_barcode) if sample_barcode else None
+        collection_barcode = sample.pop("collection_id", None) 
+        collection_identifier = find_identifier(session, collection_barcode) if collection_barcode else None
+        
+        result = {
+            "sample_barcode": sample_barcode,
+            "collection_barcode": collection_barcode
+        }
+
+        # validate barcodes
+        if sample_barcode and not sample_identifier:
+            result["status"] = "validation_failed"
+            result["details"] = f"sample barcode «{sample_barcode}» not found"
+        elif sample_identifier and sample_identifier.set_use != 'sample':
+            result["status"] = "validation_failed"
+            result["details"] = f"barcode «{sample_barcode}» has use type «{sample_identifier.set_use}»"
```

Comment:
Not a blocker, but when validating the set use (for sample or collection ids), might be nice in the error message to include the expected set_use was, e.g.

`f"sample barcode «{sample_barcode}» has use type «{sample_identifier.set_use}»"`
or
`f"barcode «{sample_barcode}» has use type «{sample_identifier.set_use}» instead of expected type «sample»"`